tripper/mappings/mappings.py

```diff
@@ -0,0 +1,618 @@
+            raise InconsistentTriplesError(
+                f"The same individual can only relate to one datamodel "
+                f"property via {instanceOf} relations."
+            )
+        soInst[s] = o
+        osInst[o].append(s)
+    soName = dict(triplestore.subject_objects(label))
+    soUnit = dict(triplestore.subject_objects(hasUnit))
+    soCost = dict(triplestore.subject_objects(hasCost))
+
+    def getcost(target, stepname):
+        """Returns the cost assigned to IRI `target` for a mapping step
+        of type `stepname`."""
+        cost = soCost.get(target, default_costs[stepname])
+        if cost is None:
+            return None
+        return function_repo[cost] if cost in function_repo else float(cost)
+
+    def walk(target, visited, step):
+        """Walk backward in rdf graph from `node` to sources."""
+        if target in visited:
+            return
+        visited.add(target)
+
+        def addnode(node, steptype, stepname):
+            if node in visited:
+                return
+            step.steptype = steptype
+            step.cost = getcost(target, stepname)
+            if node in sources:
+                value = Value(
+                    value=sources[node],
+                    unit=soUnit.get(node),
+                    iri=node,
+                    property_iri=soInst.get(node),
+                    cost=getcost(node, "value"),
+                )
+                step.add_input(value, name=soName.get(node))
+            else:
+                prevstep = MappingStep(output_iri=node, output_unit=soUnit.get(node))
+                step.add_input(prevstep, name=soName.get(node))
+                walk(node, visited, prevstep)
+
+        for node in osInst[target]:
+            addnode(node, StepType.INV_INSTANCEOF, "instanceOf")
+
+        for node in soMaps[target]:
+            addnode(node, StepType.MAPSTO, "mapsTo")
+
+        for node in osMaps[target]:
+            addnode(node, StepType.INV_MAPSTO, "mapsTo")
+
+        for node in osSubcl[target]:
+            addnode(node, StepType.INV_SUBCLASSOF, "subClassOf")
+
+        for fmap in function_mappers:
+            for func, input_iris in fmap(triplestore)[target]:
+                step.steptype = StepType.FUNCTION
+                step.cost = getcost(func, "function")
+                step.function = function_repo[func]
+                step.join_mode = True
+                for input_iri in input_iris:
+                    step0 = MappingStep(
+                        output_iri=input_iri, output_unit=soUnit.get(input_iri)
+                    )
+                    step.add_input(step0, name=soName.get(input_iri))
+                    walk(input_iri, visited, step0)
+                step.join_input()
+
+    visited = set()
+    step = MappingStep(output_iri=target, output_unit=soUnit.get(target))
+    if target in soInst:
+        # It is only initially we want to follow instanceOf in forward
+        # direction.
```

Comment:
Why? I do not understand this comment.

Reply:
Good question. It may actually be that we not only want to follow instanceOf initially. I think we need to make some drawings...